staging/src/k8s.io/apiserver/pkg/endpoints/handlers/response_test.go

```diff
@@ -222,3 +229,123 @@ func TestWatchEncoderIdentifier(t *testing.T) {
 		t.Error("  - watchEncoder::typeIdentifier to capture all relevant fields in identifier")
 	}
 }
+
+func TestWatchListEncoder(t *testing.T) {
+	scenarios := []struct {
+		name              string
+		negotiatedEncoder runtime.Serializer
+		targetGVK         *schema.GroupVersionKind
+
+		actualEvent   watch.Event
+		listBlueprint runtime.Object
+
+		expectedBase64ListBlueprint string
+	}{
+		{
+			name:              "pass through, an obj without the annotation received",
+			actualEvent:       makeBookmarkEventFor(makePod("1")),
+			negotiatedEncoder: newJsonSerializer(),
+		},
+		{
+			name:                        "encodes the initialEventsListBlueprint if an obj with the annotation is passed",
+			actualEvent:                 makeBookmarkEventFor(makePodWithInitialEventsAnnotation("1")),
+			listBlueprint:               makePodListWithKind("100"),
+			expectedBase64ListBlueprint: encodeObjectToBase64String(makePodListWithKind("100"), t),
+			negotiatedEncoder:           newJsonSerializer(),
+		},
+		{
+			name:                        "encodes the initialEventsListBlueprint as PartialObjectMetadata when requested",
+			targetGVK:                   &schema.GroupVersionKind{Group: "meta.k8s.io", Version: "v1", Kind: "PartialObjectMetadata"},
+			actualEvent:                 makeBookmarkEventFor(makePodWithInitialEventsAnnotation("2")),
+			listBlueprint:               makePodListWithKind("101"),
+			expectedBase64ListBlueprint: encodeObjectToBase64String(makePartialObjectMetadataListWithoutKind("101"), t),
+			negotiatedEncoder:           newJsonSerializer(),
+		},
+	}
+
+	for _, scenario := range scenarios {
+		t.Run(scenario.name, func(t *testing.T) {
+			target := newWatchListTransformer(scenario.listBlueprint, scenario.targetGVK, scenario.negotiatedEncoder)
+			transformedEvent := target.transform(scenario.actualEvent)
+
+			actualObjectMeta, err := meta.Accessor(transformedEvent.Object)
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			base64ListBlueprint, ok := actualObjectMeta.GetAnnotations()[metav1.InitialEventsListBlueprintAnnotationKey]
+			if !ok && len(scenario.expectedBase64ListBlueprint) != 0 {
+				t.Fatalf("the encoded obj doesn't have %q", metav1.InitialEventsListBlueprintAnnotationKey)
+			}
+			if base64ListBlueprint != scenario.expectedBase64ListBlueprint {
+				t.Fatalf("unexpected base64ListBlueprint = %s, expected = %s", base64ListBlueprint, scenario.expectedBase64ListBlueprint)
+			}
+		})
+	}
+}
+
+func makePartialObjectMetadataListWithoutKind(rv string) *metav1.PartialObjectMetadataList {
+	return &metav1.PartialObjectMetadataList{
+		// do not set the type info to match
+		// newWatchListTransformer
```

Comment:
because `asPartialObjectMetadataList` creates `PartialObjectMetadataList` without the kind (https://github.com/kubernetes/kubernetes/pull/127587/files#diff-12e0758457373aa860bb0baae0878a99c107840d25fcf356d126d4b3d1d15663R569)